File: crates/flighthq-camera/src/picking.rs
//! Screen-to-world ray unprojection and world-to-screen projection.
//!
//! Port of `picking.ts` from `@flighthq/camera`.

use flighthq_geometry::{inverse_matrix4, normalize_vector3, subtract_vector3};
use flighthq_types::{Camera, Matrix4Like, Ray3D, Vector3Like};

use crate::camera::get_camera_view_projection_matrix4;

/// Writes the world-space ray from the camera through an NDC screen point
/// into `out` and returns `true`. Returns `false` (leaving `out` untouched)
/// when the view-projection is non-invertible.
///
/// `ndc_x` / `ndc_y` are normalized device coordinates in `[-1, 1]`
/// (center = 0, top-right = (1, 1)). `aspect` is viewport width / height.
///
/// The returned ray's origin is the near-plane world position and its
/// direction is the normalized world-space direction from near through far.
///
/// Alias-safe: reads all inputs before writing `out`.
pub fn get_camera_screen_to_world_ray(
    out: &mut Ray3D,
    camera: &Camera,
    ndc_x: f32,
    ndc_y: f32,
    aspect: f32,
) -> bool {
    let mut vp = Matrix4Like::default();
    let mut inv_vp = Matrix4Like::default();
    get_camera_view_projection_matrix4(&mut vp, camera, aspect);
    if !inverse_matrix4(&mut inv_vp, &vp) {
        return false;
    }
    let m = &inv_vp.m;
    let nx = ndc_x;
    let ny = ndc_y;

    // Unproject near-plane point (nx, ny, -1, 1)
    let mut near_x = m[0] * nx + m[4] * ny + m[8] * -1.0 + m[12];
    let mut near_y = m[1] * nx + m[5] * ny + m[9] * -1.0 + m[13];
    let mut near_z = m[2] * nx + m[6] * ny + m[10] * -1.0 + m[14];
    let near_w = m[3] * nx + m[7] * ny + m[11] * -1.0 + m[15];
    if near_w != 0.0 {
        let inv_w = 1.0 / near_w;
        near_x *= inv_w;
        near_y *= inv_w;
        near_z *= inv_w;
    }

    // Unproject far-plane point (nx, ny, 1, 1)
    let mut far_x = m[0] * nx + m[4] * ny + m[8] + m[12];
    let mut far_y = m[1] * nx + m[5] * ny + m[9] + m[13];
    let mut far_z = m[2] * nx + m[6] * ny + m[10] + m[14];
    let far_w = m[3] * nx + m[7] * ny + m[11] + m[15];
    if far_w != 0.0 {
        let inv_w = 1.0 / far_w;
        far_x *= inv_w;
        far_y *= inv_w;
        far_z *= inv_w;
    }

    // Direction: from near to far, normalized. Stack scratch values.
    let near_like = Vector3Like {
        x: near_x,
        y: near_y,
        z: near_z,
    };
    let far_like = Vector3Like {
        x: far_x,
        y: far_y,
        z: far_z,
    };
    let mut dir = Vector3Like::default();
    subtract_vector3(&mut dir, &far_like, &near_like);
    let dir_copy = Vector3Like {
        x: dir.x,
        y: dir.y,
        z: dir.z,
    };
    normalize_vector3(&mut dir, &dir_copy);

    out.origin.x = near_x;
    out.origin.y = near_y;
    out.origin.z = near_z;
    out.direction.x = dir.x;
    out.direction.y = dir.y;
    out.direction.z = dir.z;
    true
}

/// Writes the NDC coordinates `(x, y in [-1, 1], z = clip-space depth)` of a
/// world-space point into `out`. Returns `false` (leaving `out` untouched) when
/// the point is behind the camera (`w <= 0`). `aspect` is viewport width / height.
///
/// Alias-safe: reads all inputs before writing `out`.
pub fn get_camera_world_to_screen(
    out: &mut Vector3Like,
    camera: &Camera,
    world_point: &Vector3Like,
    aspect: f32,
) -> bool {
    let mut vp = Matrix4Like::default();
    get_camera_view_projection_matrix4(&mut vp, camera, aspect);
    let m = &vp.m;
    let wx = world_point.x;
    let wy = world_point.y;
    let wz = world_point.z;
    let clip_x = m[0] * wx + m[4] * wy + m[8] * wz + m[12];
    let clip_y = m[1] * wx + m[5] * wy + m[9] * wz + m[13];
    let clip_z = m[2] * wx + m[6] * wy + m[10] * wz + m[14];
    let clip_w = m[3] * wx + m[7] * wy + m[11] * wz + m[15];
    if clip_w <= 0.0 {
        return false;
    }
    let inv_w = 1.0 / clip_w;
    out.x = clip_x * inv_w;
    out.y = clip_y * inv_w;
    out.z = clip_z * inv_w;
    true
}

#[cfg(test)]
mod tests {
    use flighthq_geometry::create_vector3;

    use super::*;
    use crate::camera::{create_camera, set_camera_view_matrix4_from_look_at};
    use crate::projection::create_perspective_projection;

    fn make_camera() -> Camera {
        let projection = create_perspective_projection(std::f32::consts::FRAC_PI_3, 1.0);
        let mut camera = create_camera(0.1, 100.0, projection);
        set_camera_view_matrix4_from_look_at(
            &mut camera,
            &create_vector3(0.0, 0.0, 5.0),
            &create_vector3(0.0, 0.0, 0.0),
            &create_vector3(0.0, 1.0, 0.0),
        );
        camera
    }

    mod get_camera_screen_to_world_ray {
        use super::*;

        #[test]
        fn center_ndc_origin_z_is_less_than_far() {
            let camera = make_camera();
            let mut ray = Ray3D::default();
            let ok = get_camera_screen_to_world_ray(&mut ray, &camera, 0.0, 0.0, 1.0);
            assert!(ok, "expected invertible view-projection");
            assert!(
                ray.origin.z < camera.far,
                "origin.z ({}) should be less than far ({})",
                ray.origin.z,
                camera.far
            );
        }

        #[test]
        fn direction_is_unit_length() {
            let camera = make_camera();
            let mut ray = Ray3D::default();
            assert!(get_camera_screen_to_world_ray(
                &mut ray, &camera, 0.0, 0.0, 1.0
            ));
            let len = (ray.direction.x * ray.direction.x
                + ray.direction.y * ray.direction.y
                + ray.direction.z * ray.direction.z)
                .sqrt();
            assert!(
                (len - 1.0).abs() < 1e-5,
                "direction not normalized: len={len}"
            );
        }
    }

    mod get_camera_world_to_screen {
        use super::*;

        #[test]
        fn look_at_target_projects_to_ndc_center() {
            let camera = make_camera();
            // The look-at target (origin) should project to approximately NDC (0, 0).
            let world = Vector3Like {
                x: 0.0,
                y: 0.0,
                z: 0.0,
            };
            let mut ndc = Vector3Like::default();
            assert!(get_camera_world_to_screen(&mut ndc, &camera, &world, 1.0));
            assert!((ndc.x).abs() < 1e-4, "expected ndc.x near 0, got {}", ndc.x);
            assert!((ndc.y).abs() < 1e-4, "expected ndc.y near 0, got {}", ndc.y);
        }

        #[test]
        fn returns_false_for_point_behind_camera() {
            let camera = make_camera();
            // Eye is at z=5; a point at z=200 is behind the camera.
            let world = Vector3Like {
                x: 0.0,
                y: 0.0,
                z: 200.0,
            };
            let mut ndc = Vector3Like::default();
            assert!(
                !get_camera_world_to_screen(&mut ndc, &camera, &world, 1.0),
                "expected false for point behind camera"
            );
        }
    }
}

File: packages/camera/src/picking.ts
import { createMatrix4, createVector3, inverseMatrix4, normalizeVector3, subtractVector3 } from '@flighthq/geometry';
import type { Camera3D, Ray3DLike, Vector3Like } from '@flighthq/types';

import { getCamera3DViewProjectionMatrix4 } from './camera';

// Writes the world-space ray from the camera through an NDC screen point (ndcX, ndcY) into
// `out` and returns true, or returns false (leaving `out` untouched) when the view-projection
// is non-invertible. NDC coordinates range from -1 to +1 on both axes (center = 0, top-right =
// (1, 1)). `aspect` is viewport width / height.
//
// The returned ray origin is the near-plane world position and the direction is the normalized
// world-space direction from near through far. Use this for mouse picking, drag raycasts, and
// any screen-to-world query.
//
// Reads all inputs into locals before writing `out`, so it is alias-safe.
export function getCamera3DScreenToWorldRay(
  out: Ray3DLike,
  camera: Readonly<Camera3D>,
  ndcX: number,
  ndcY: number,
  aspect: number,
): boolean {
  getCamera3DViewProjectionMatrix4(__scratchViewProjection, camera, aspect);
  if (!inverseMatrix4(__scratchInverseVP, __scratchViewProjection)) {
    return false;
  }
  const m = __scratchInverseVP.m;
  const nx = ndcX,
    ny = ndcY;
  // Unproject near-plane point (ndcX, ndcY, -1, 1) through the inverse view-projection.
  let nearX = m[0] * nx + m[4] * ny + m[8] * -1 + m[12];
  let nearY = m[1] * nx + m[5] * ny + m[9] * -1 + m[13];
  let nearZ = m[2] * nx + m[6] * ny + m[10] * -1 + m[14];
  const nearW = m[3] * nx + m[7] * ny + m[11] * -1 + m[15];
  if (nearW !== 0) {
    const invW = 1 / nearW;
    nearX *= invW;
    nearY *= invW;
    nearZ *= invW;
  }
  // Unproject far-plane point (ndcX, ndcY, 1, 1).
  let farX = m[0] * nx + m[4] * ny + m[8] + m[12];
  let farY = m[1] * nx + m[5] * ny + m[9] + m[13];
  let farZ = m[2] * nx + m[6] * ny + m[10] + m[14];
  const farW = m[3] * nx + m[7] * ny + m[11] + m[15];
  if (farW !== 0) {
    const invW = 1 / farW;
    farX *= invW;
    farY *= invW;
    farZ *= invW;
  }
  // Direction: from near to far, normalized.
  __scratchNear.x = nearX;
  __scratchNear.y = nearY;
  __scratchNear.z = nearZ;
  __scratchFar.x = farX;
  __scratchFar.y = farY;
  __scratchFar.z = farZ;
  subtractVector3(__scratchDir, __scratchFar, __scratchNear);
  normalizeVector3(__scratchDir, __scratchDir);
  out.origin.x = nearX;
  out.origin.y = nearY;
  out.origin.z = nearZ;
  out.direction.x = __scratchDir.x;
  out.direction.y = __scratchDir.y;
  out.direction.z = __scratchDir.z;
  return true;
}

// Writes the NDC coordinates (x, y ∈ [-1, 1], z = clip-space depth) of a world-space point
// into `out` and returns true, or returns false (leaving `out` untouched) when the point is
// behind the camera (w <= 0) or when the perspective divide yields a degenerate result.
// `aspect` is viewport width / height.
//
// Use this to anchor HUD elements, labels, and gizmos to world positions.
//
// Reads all inputs into locals before writing `out`, so it is alias-safe.
export function getCamera3DWorldToScreen(
  out: Vector3Like,
  camera: Readonly<Camera3D>,
  worldPoint: Readonly<Vector3Like>,
  aspect: number,
): boolean {
  getCamera3DViewProjectionMatrix4(__scratchViewProjection, camera, aspect);
  const m = __scratchViewProjection.m;
  const wx = worldPoint.x,
    wy = worldPoint.y,
    wz = worldPoint.z;
  const clipX = m[0] * wx + m[4] * wy + m[8] * wz + m[12];
  const clipY = m[1] * wx + m[5] * wy + m[9] * wz + m[13];
  const clipZ = m[2] * wx + m[6] * wy + m[10] * wz + m[14];
  const clipW = m[3] * wx + m[7] * wy + m[11] * wz + m[15];
  if (clipW <= 0) {
    return false;
  }
  const invW = 1 / clipW;
  out.x = clipX * invW;
  out.y = clipY * invW;
  out.z = clipZ * invW;
  return true;
}

// Scratch objects reused across calls. Single-threaded; not re-entrant.
const __scratchViewProjection = createMatrix4();
const __scratchInverseVP = createMatrix4();
const __scratchNear = createVector3();
const __scratchFar = createVector3();
const __scratchDir = createVector3();
